### PR TITLE
Add Error Handling For Libraries

### PR DESCRIPTION
**Pull request by Gianfranco99**
This resolve [issue #23](https://github.com/lambdaclass/yet-another-swap/issues/23). With the recommendations in[ the "Cairo Book" for error handling](https://book.cairo-lang.org/ch10-00-error-handling.html), replacing assert statements with the Result enum in the specified libraries.

### CODE REVIEW

File: crates/yas_core/src/libraries/bit_math.cairo [hunk 1]
```diff
@@ -98,4 +108,12 @@ mod BitMath {
         }
         r
     }
+
+    fn check_value(x: u256) -> Result<(), felt252> {
```
Comment (dubzn) [inline crates/yas_core/src/libraries/bit_math.cairo:112]:
```suggestion
    fn check_gt_zero(x: u256) -> Result<(), felt252> {
```

File: crates/yas_core/src/libraries/liquidity_math.cairo [hunk 2]
```diff
@@ -15,14 +15,39 @@ mod LiquidityMath {
...
+                Result::Ok(()) => {},
+                Result::Err(err) => {
+                    panic_with_felt252(err)
+                }
+            };
             x + y_u128
         }
     }
+
+    fn check_value(x: u128, y: u128) -> Result<(), felt252> {
```
Comment (dubzn) [inline crates/yas_core/src/libraries/liquidity_math.cairo:39]:
```suggestion
    fn check_liquidity_sub(x: u128, y: u128) -> Result<(), felt252> {
```

File: crates/yas_core/src/libraries/liquidity_math.cairo [hunk 3]
```diff
@@ -15,14 +15,39 @@ mod LiquidityMath {
...
+        if x >= y {
+            Result::Ok(())
+        } else {
+            Result::Err('LS')
+        }
+    }
+    fn check_overflowing_add(x: u128, y: u128) -> Result<(), felt252> {
```
Comment (dubzn) [inline crates/yas_core/src/libraries/liquidity_math.cairo:46]:
```suggestion
    fn check_liquidity_add(x: u128, y: u128) -> Result<(), felt252> {
```

File: crates/yas_core/src/tests/test_libraries/test_sqrt_price_math.cairo [hunk 4]
```diff
@@ -18,7 +18,8 @@ mod TestSqrtPriceMath {
 
         // fails if price is zero
         #[test]
-        #[should_panic]
+        #[available_gas(20000000)]
+        #[should_panic(expected: ('sqrt_ratio_AX96 cannot be neg',))]
```
Comment (dubzn) [inline crates/yas_core/src/tests/test_libraries/test_sqrt_price_math.cairo:22]: :fire:

File: crates/yas_core/src/libraries/tick_math.cairo [hunk 5]
```diff
@@ -35,9 +35,12 @@ mod TickMath {
...
         let abs_tick = tick.abs();
-        assert(
-            abs_tick <= MAX_TICK(), 'T'
-        ); // TODO: review this error in the future. This is the original error from UniswapV3.
+        match check_abs_tick(abs_tick) {
+            Result::Ok(()) => {},
+            Result::Err(err) => {
+                panic_with_felt252(err)
+            }
+        }; // TODO: review this error in the future. This is the original error from UniswapV3.
```
Comment (dubzn) [inline crates/yas_core/src/libraries/tick_math.cairo:43]:
```suggestion
        }; 
```

File: crates/yas_core/src/libraries/tick_math.cairo [hunk 6]
```diff
@@ -299,4 +303,21 @@ mod TickMath {
...
+    fn check_sqrt_priceX96(sqrt_priceX96: FixedType) -> Result<(), felt252> {
+        if sqrt_priceX96 >= FixedTrait::new(MIN_SQRT_RATIO, false)
+            && sqrt_priceX96 < FixedTrait::new(MAX_SQRT_RATIO, false) {
+            Result::Ok(())
+        } else {
+            Result::Err('R')
+        }
+    }
+
+    fn check_abs_tick(tick: i32) -> Result<(), felt252> {
```
Comment (dubzn) [inline crates/yas_core/src/libraries/tick_math.cairo:316]:
```suggestion
    // T: The given tick must be less than, or equal to, the maximum tick
    fn check_tick(tick: i32) -> Result<(), felt252> {
```

**Review by dubzn — CHANGES_REQUESTED** [review]
GJ, very nice! I'll leave some brief comments related to the function names. I'd also like to discuss whether we want to name them `check_x` or `assert_x`.

References: [Uniswap Error Codes](https://docs.uniswap.org/contracts/v3/reference/error-codes)

File: crates/yas_core/src/libraries/bit_math.cairo (hunk 1 above)
Comment (uri-99) [reply crates/yas_core/src/libraries/bit_math.cairo:112]: +1

File: crates/yas_core/src/libraries/liquidity_math.cairo (hunk 2 above)
Comment (uri-99) [reply crates/yas_core/src/libraries/liquidity_math.cairo:39]: +1

File: crates/yas_core/src/libraries/liquidity_math.cairo (hunk 3 above)
Comment (uri-99) [reply crates/yas_core/src/libraries/liquidity_math.cairo:46]: +1

File: crates/yas_core/src/libraries/sqrt_price_math.cairo [hunk 7]
```diff
@@ -45,8 +49,18 @@ mod SqrtPriceMath {
         } else {
             // if the product overflows, we know the denominator underflows
             // in addition, we must check that the denominator does not underflow
-            assert(FP64x96Impl::new(product / amount, false) == sqrtPX96, 'product overflow');
-            assert(numerator > product, 'denominator underflow');
+            match check_product_overflow(FP64x96Impl::new(product / amount, false), sqrtPX96) {
```
Comment (uri-99) [inline crates/yas_core/src/libraries/sqrt_price_math.cairo:52]:
```suggestion
            match check_product_overflow(product, amount, false, sqrtPX96) {
```

File: crates/yas_core/src/libraries/sqrt_price_math.cairo [hunk 8]
```diff
@@ -203,4 +242,45 @@ mod SqrtPriceMath {
...
+        if sqrtPX96 > quotient {
+            Result::Ok(())
+        } else {
+            Result::Err('sqrtPX96_fp < quotient')
+        }
+    }
+
+    fn check_product_overflow(value_1: FixedType, value_2: FixedType) -> Result<(), felt252> {
+        if value_1 == value_2 {
+            Result::Ok(())
```
Comment (uri-99) [inline crates/yas_core/src/libraries/sqrt_price_math.cairo:273]:
```suggestion
    fn check_product_overflow(product: u256, amount: u256, sign: bool, value_2: FixedType) -> Result<(), felt252> {
        if FP64x96Impl::new(product / amount, sign) == value_2 {
            Result::Ok(())
```

**Review by uri-99 — CHANGES_REQUESTED** [review]
Here, the process of the check should be done inside the check_product_overflow function, otherwise, check_product_overflow is just a check_equals

File: crates/yas_core/src/libraries/sqrt_price_math.cairo [hunk 9]
```diff
@@ -203,4 +242,45 @@ mod SqrtPriceMath {
             )
         }
     }
+
+    fn check_sqrtPX96_sign(sign: bool) -> Result<(), felt252> {
+        if sign == false {
+            Result::Ok(())
+        } else {
+            Result::Err('sqrt_ratio_AX96 cannot be neg')
+        }
```
Comment (uri-99) [inline crates/yas_core/src/libraries/sqrt_price_math.cairo:251]: Every operation costs gas, so we should minimize unnecessary operations to save on gas.
```suggestion
        if sign {
            Result::Err('sqrt_ratio_AX96 cannot be neg')
        } else {
            Result::Ok(())
        }
```

**Review by uri-99 — CHANGES_REQUESTED** [review]
Good job! I left just some small suggestions to improve the style of coding, for efficiency and comprehension. 👍

**Review by uri-99 — APPROVED** [review]

**Review by dubzn — DISMISSED** [review]

**Review by dpinones — APPROVED** [review]

**Pull request merged**